File: livingrimoire start here/LivinGrimoire java/LivinGrimoire 030123/LivinGrimoire core/Cerabellum.java
package LivinGrimoire;

public class Cerabellum {
    // runs an algorithm
    private int fin;
    private int at;
    private enumFail failType;
    private Boolean incrementAt = false;

    public void advanceInAlg() {
        if (incrementAt) {
            incrementAt = false;
            at++;
            if (at == fin) {
                isActive = false;
            }
        }
    }

    public int getAt() {
        return at;
    }

    public Algorithm alg;
    private boolean isActive = false;
    private String emot = "";

    public String getEmot() {
        return emot;
    }
    public boolean setAlgorithm(Algorithm algorithm) {
        if (!isActive && (!algorithm.getAlgParts().isEmpty())) {
            this.alg = algorithm;
            this.at = 0;
            this.fin = algorithm.getSize();
            this.isActive = true;
            this.emot = alg.getAlgParts().get(at).myName(); // updated line
            return false;
        }
        return true;
    }

    public boolean isActive() {
        return isActive;
    }

    public boolean setActive(Boolean b1) {
        return isActive = b1;
    }
    public void setActive(boolean isActive) {
        this.isActive = isActive;
    }

    public String act(String ear, String skin, String eye) {
        String axnStr = "";
        if (!isActive) {
            return axnStr;
        }
        if (at < fin) {
            axnStr = alg.getAlgParts().get(at).action(ear, skin, eye);
            this.emot = alg.getAlgParts().get(at).myName();
            if (alg.getAlgParts().get(at).completed()) {
                incrementAt = true;
            }
        }
        return axnStr;
    }

    public int getMutationLimitOfActiveAlgPart() {
        return alg.getAlgParts().get(at).getMutationLimit();
    }
    public enumFail getFailType() {
        return alg.getAlgParts().get(at).failure("");
    }
}

File: livingrimoire start here/LivingGrimoire Kotlin/lgkt/Cerabellum.kt
package lgkt

open class Cerabellum {
    // runs an algorithm
    private var fin = 0
    var at = 0
        private set
    val failType: enumFail?
        get() = alg!!.algParts[at].failure("")
    private var incrementAt = false
    fun advanceInAlg() {
        if (incrementAt) {
            incrementAt = false
            at++
            if (at == fin) {
                isActive = false
            }
        }
    }

    var alg: Algorithm? = null
    var isActive = false
    var emot = ""
        private set

    fun setAlgorithm(algorithm: Algorithm): Boolean {
        if (!isActive && !algorithm.algParts.isEmpty()) {
            alg = algorithm
            at = 0
            fin = algorithm.size
            isActive = true
            emot = alg!!.algParts[at].myName() // updated line
            return false
        }
        return true
    }

    fun setActive(b1: Boolean): Boolean {
        return b1.also { isActive = it }
    }

    open fun act(ear: String, skin: String, eye: String): String {
        var axnStr = ""
        if (!isActive) {
            return axnStr
        }
        if (at < fin) {
            axnStr = alg!!.algParts[at].action(ear, skin, eye)
            emot = alg!!.algParts[at].myName()
            if (alg!!.algParts[at].completed()) {
                incrementAt = true
            }
        }
        return axnStr
    }

    val mutationLimitOfActiveAlgPart: Int
        get() = alg!!.algParts[at].getMutationLimit()
}